Local allocation prototype: initial type system plumbing and backend support

stedolan opened the pull request:
This is the start of a prototype to add escape analysis and stack allocation via modal types to OCaml. 

This first patch is mostly plumbing. There's a couple of FIXMEs in the code, some of which we should figure out during review and some of which should wait for a later patch.

The features in this patch are:

### Typing
  - A new exp_mode field in Typedtree nodes
  - Enough plumbing of mode values through Typecore to fill in exp_mode with stub values
  - Modes and locking in typing environments (although without much of interest to do with them)

### Lambda
  - New region operations `Lbeginregion` and `Pendregion`. Note the asymmetry: starting a region is a block-structured operation. It may be ended at any point with the `Pendregion` primitive. A single `Pendregion` may end multiple regions, as ending an outer region implicitly ends inner regions.
  - Automatic insertion of `Lbeginregion` and `Pendregion` whenever we transition into the stack-allocation mode.
  - Annotations of the `Pmakeblock` primitive with the appropriate allocation mode.

### Backend
  - Plumbing of the new operations through Closure / Clambda / Cmm. By Cmm time, the block structure is gone and the region operations are lowered to external C calls.
  - (Flambda compiles but is untested)

### Assembly / Runtime
  - Basic runtime implementation for amd64 only.
  - External C calls for region manipulation (easy to break on in gdb, but will not perform well)

---

Conspicuously missing (for future patches):

  - Interesting typechecking: the only way to make a stack allocation is `let[@stack]`, which is special-cased in the typechecker and is not checked for soundness.
  - GC support: runtime support so far extends only to performing stack allocations. They're not currently traced during GC, so crashes will occur if they are live across a collection.

stedolan committed:
Initial version of local allocation (unsafe)

stedolan committed:
More typing modes and locking of environments

lpw25 commented on typing/types.mli:
```diff
@@ -584,3 +584,18 @@ type label_description =
+  val max_mode : t
+  val is_max : t -> bool
+
+  val constrain : t -> t -> (unit, unit) result
```
Maybe a name that indicates the nature of the constraint? Less_than? Subkind?

stedolan replied:
Reasonable, but I'll merge this as 'constrain' and change it in #21 instead. (It's used a lot more there and conflicts will be annoying otherwise)

lpw25 commented on typing/typecore.ml:
```diff
@@ -2642,12 +2655,12 @@ and type_expect_
         exp_desc = Texp_constant cst;
         exp_loc = loc; exp_extra = [];
-        exp_type = instance Predef.type_string;
+        exp_type = instance Predef.type_string; exp_mode = Alloc_heap; (*FIXME*)
```
Out of interest, why not just use `mode` here already? Is it just a lack of support for local strings?

stedolan replied:
At the time, I was unsure which mode was appropriate here. Constants are always shared, never local. I think exp_mode=mode is more appropriate, though.

lpw25 commented on typing/typecore.ml:
```diff
@@ -3732,7 +3751,7 @@ and type_ident env ?(recarg=Rejected) lid =
 and type_binding_op_ident env s =
```
This should probably take a `mode` parameter too.

stedolan replied:
Probably! I left out the binding operator stuff mostly because I didn't understand this function.

stedolan replied:
On second thoughts, I'm going to leave this as-is for now. Getting binding operators to work with local allocations will be somewhat tricky, and involves some design work that seems better left for later.

lpw25 replied:
That's fine, but can you leave a fixme comment in to make sure we don't forget.

lpw25 commented on typing/env.ml:
```diff
@@ -221,8 +221,8 @@ module IdTbl =
     type ('a, 'b) t = {
-      current: 'a Ident.tbl;
-      (** Local bindings since the last open *)
+      current: (alloc_mode * 'a) Ident.tbl;
```
It seems a shame to keep a mode on all bindings in the environment when it only makes sense for values (and maybe modules and classes later). I think you could instead have the various access functions take a `lock` parameter of type `'a -> lock -> 'a` and then put the `alloc_mode` in the `'a` itself (i.e. `value_entry`). Alternatively, you could implement locking using the existing `Map` constructor.

stedolan replied:
I thought about this a bit, and I think the cleanest option is to make alloc_mode be a type parameter here. (I don't fully follow the suggestion about `Map`).

lpw25 replied:
`Map` let's you add an arbitrary function that is mapped on all values the are looked up through that `Map`. This is pretty similar to the definition of a lock. If you have a lock as part of `value_entry` (which would be just an "empty" lock initially) then you can use something like:
```ocaml
IdTbl.map (function
  | Val_unbound _ as desc -> desc
  | Val_bound(desc, lock) -> Val_bound(desc, add_lock new_lock lock)) env
```
to add `new_lock` to `env`.

lpw25 replied:
You most recent version looks pretty good. But I think it would be simpler if the type variable represented just the locks not the locks and the original mode. I think you could then push the original mode into `value_entry` or `value_data`.

lpw25 commented on typing/env.ml:
```diff
@@ -273,53 +281,65 @@ module IdTbl =
+              find_name_and_modes wrap ~mark name next macc in
+            p, m, macc, f desc
+        | Lock {mode; next} ->
+            find_name_and_modes wrap ~mark name next (mode :: macc)
```
Why accumulate the modes rather than reduce them immediately?

stedolan replied:
To keep the error-reporting logic with the lookup code, rather than mixing it into IdTbl.

lpw25 replied:
It looks to me like good error-reporting only requires that we keep the original variable mode separate from the lock. We should still be able to collapse the locks to a single greatest (least?) one.

stedolan replied:
By introducing new variables? That seems worse than making a list.

lpw25 replied:
Discussed offline: I thought that these would be lists of mostly constants, whereas they will probably be lists of variables, which means they can't meaningfully be reduced.

lpw25 commented on lambda/lambda.mli:
```diff
@@ -288,6 +294,7 @@ type lambda =
   | Lsend of meth_kind * lambda * lambda * lambda list * scoped_location
   | Levent of lambda * lambda_event
   | Lifused of Ident.t * lambda
+  | Lbeginregion of Ident.t * lambda
```
Does this need to be in `lambda` rather than `primitive`? What are the semantics of:
```ocaml
Lbeginregion (id, Lsequence[e1; e2])
```
vs.
```ocaml
Llet(id, (Lbeginregion(id, Lseqeunce [e1; Lvar id]), e2)
```

stedolan replied:
The second is invalid (and possibly the first too). The invariant (that should be documented in lambda.mli) is that Pendregion must be called on the variable bound by Lbeginregion before the Lbeginregion ends, and there should be no other uses of the region variable.

lpw25 commented on lambda/translcore.ml:
```diff
@@ -218,6 +223,58 @@ let transl_ident loc env ty path desc =
+       let sw =
+         {sw with
+           sw_consts = List.map (fun (i,l) -> i, endregion l) sw.sw_consts;
+           sw_blocks = List.map (fun (i,l) -> i, endregion l) sw.sw_blocks } in
```
Missing `sw_failaction`

stedolan replied:
Good catch.

lpw25 commented on lambda/translcore.ml:
```diff
@@ -218,6 +223,58 @@ let transl_ident loc env ty path desc =
+  let res = Ident.create_local "res" in
+  let op_end =
+    Lprim (Pendregion, [Lvar r], Debuginfo.Scoped_location.Loc_unknown) in
+  let rec endregion = function
```
I wonder if we should be trying to avoid duplicating the endregion down every branch if we can.

I think that, once all the cases are filled in, we will not be performing the endregion down some branches, since a branch that ends in a tail-call with local parameters should not end the region. But even then we can still share the endregion using staticraise/staticcatch.

lpw25 commented on lambda/lambda.mli:
```diff
@@ -56,7 +60,7 @@ type primitive =
   | Pgetglobal of Ident.t
   | Psetglobal of Ident.t
   (* Operations on heap blocks *)
-  | Pmakeblock of int * mutable_flag * block_shape
```
This isn't the only primitive that produces allocations. Might worth adding the rest as part of this PR.

lpw25 commented on lambda/translcore.ml:
```diff
@@ -738,11 +812,13 @@ and transl_apply ~scopes
+                            Btype.is_optional l)
+                          sargs)
+  in
+  maybe_region mode bound_modes lam
```
What are the criteria for having `maybe_region` on let and apply but not on any other constructs? It seems to make some sense, but is there a concrete property they have that other constructs don't.

stedolan replied:
You need a `maybe_region` whenever there is an expression in mode `Alloc_shared` that evaluates a subexpression in mode `Alloc_local`, because only then is there a possibility of a local allocation that must be deleted.

This can only happen on a handful of expressions. Let and apply are the most important cases, but there are a couple of others currently missing (Match, in particular)

lpw25 commented on middle_end/flambda/un_anf.ml:
```diff
@@ -447,6 +449,9 @@ let let_bound_vars_that_can_be_moved var_info (clam : Clambda.ulambda) =
       let_stack := []
+    | Ubeginregion (_r, e) ->
+      let_stack := [];
+      loop e  (* FIXME is this correct? *)
```
I think this is correct

lpw25 commented on middle_end/semantics_of_primitives.ml:
```diff
@@ -133,6 +133,7 @@ let for_primitive (prim : Clambda_primitives.primitive) =
       (* Removed by [Closure_conversion] in the flambda pipeline. *)
       No_effects, No_coeffects
+  | Pendregion -> Arbitrary_effects, Has_coeffects
 
```
I think you may also need to change the semantics of `Pmakeblock` when it is `Alloc_local` to include co-effects. Otherwise I think flambda can in principle decide to move these allocations relative to the regions.

stedolan replied:
Good point.

lpw25 commented on asmcomp/amd64/CSE.ml:
```diff
@@ -29,6 +29,8 @@ method! class_of_operation op =
     begin match spec with
     | Ilea _ | Isextend32 | Izextend32 -> Op_pure
     | Istore_int(_, _, is_asg) -> Op_store is_asg
+    | Iregionbegin -> Op_load
```
I think `Op_load` is a little risky here. We need this value to be cleared by stack allocations, but in principle `Op_load` should not be affected by stack allocations (under the assumption that stack allocations never allow the code to get interrupted -- which I'm not sure whether we've decided on).

stedolan replied:
Yes. I've changed both to Op_other: these should just not participate in CSE.

lpw25 commented on asmcomp/amd64/emit.mlp:
```diff
@@ -649,6 +649,20 @@ let emit_instr fallthrough i =
+      I.cmp (domain_field Domainstate.Domain_local_limit) r;
+      let lbl_ok = new_label () in
+      I.j GE (label lbl_ok);
+      emit_call "caml_call_local_realloc";
```
Presumably this should be outlined like the calls to `caml_call_gc`.

lpw25 reviewed (commented):
Looks good. A few minor comments.

stedolan committed:
amd64: Move stack realloc calls out-of-line

stedolan committed:
Retain block-structured local regions until Mach.

Insertion of explicit endregion primitives now occurs in Selectgen,
rather than in Lambda. Early exit from regions (via static exceptions)
is now permitted, with endregions inserted where appropriate.

stedolan commented:
I've pushed a new version that preserved block-structure for longer. Lambda now contains a single new block construct `Lregion e`, and it is now valid to exit early from `Lregion` using static exceptions. (The begin/end primitives are now inserted at Selectgen time)

lpw25 commented on asmcomp/selectgen.ml:
```diff
@@ -60,8 +65,24 @@ let env_find_static_exception id env =
+(* Assuming [regions] is equal to or a suffix of [env.regions],
+   return the last region in [env.regions] but not [regions]
+   (or None if they are equal) *)
+let env_close_regions env rs =
```
s/rs/regions/ to match comment (or vice-versa)

lpw25 commented on asmcomp/selectgen.ml:
```diff
@@ -780,6 +795,9 @@ method emit_expr (env:environment) exp =
+      let env =
+        (* Disable region-fusion on loops *)
+        match rec_flag with Recursive -> env' | Nonrecursive -> env in
       let env =
```
Shouldn't the body always be non-tail? For cases like:
```
(region
   (catch
      (region (...; exit 0))
    with exit 0 ->
      ...))
```

stedolan replied:
This is intended to catch cases like:
```
(region
  ...
  (catch
     (if ...
       (exit 0)
       (region E))
    with exit 0 -> ...)))
```
where `E` contains no other `exit 0` occurrences. This is a very common code pattern due to how pattern-matching is compiled, and it is useful to fuse regions in this case.

However, this will extend some lifetimes if code like your example occurs. I expect this to be very rare, so I'm tempted to leave it as is. (Such extension of lifetimes will also be caused by the simplify_exits optimisation in Simplif, I think)

I think a more sophisticated check might be to allow fusion of regions outside and in the body of a catch, if the inner region contains no references to that catch's handlers.

lpw25 replied:
I think it would be worth doing the more sophisticated check if it can be done easily. Similarly, I think it would be worth fixing `simplify_exits`.

stedolan replied:
I have a fix for simplify_exits that I'll post upstream soon. I don't see a particularly easy way to do the more sophisticated check here.

lpw25 reviewed (commented):
The new code looks good.

lpw25 reviewed (approved):
Looks good to me now.
